include/zephyr/storage/storage_area/storage_area_flash.h

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for storage area flash subsystem
+ */
+
+#ifndef ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_FLASH_H_
+#define ZEPHYR_INCLUDE_STORAGE_STORAGE_AREA_FLASH_H_
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/storage/storage_area/storage_area.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define STORAGE_AREA_FLASH_NO_XIP (-1)
+/**
+ * @brief Storage_area_flash interface
+ * @defgroup storage_area_flash_interface Storage_area_flash interface
+ * @ingroup storage_apis
+ * @{
+ */
+
+struct storage_area_flash {
+	const struct storage_area area;
+	const struct device *dev;
+	const size_t start;
+	uintptr_t xip_address;
```

Comment:
This is no-go in this PR, I understand what it is for, because also use something like this to  directly calculate sha on device in MCUboot, and skip on flash reads, therefore I see the potential in this for optimizing  read functions in many instances.

The problem is that currently, in this PR this is not tested nor there is any actual implementation around it.
Which means that this, nor the usage, has been properly evaluated against hardware, therefore there is no proper way to advertise this to users.

Example here is that internal devices may work fine (with some exceptions), but it is currently rare for XIP to have awareness of writes that happened and therefore lack cache invalidation,  in short: write to area that is already cached by read may not be reflected by consecutive read from the same area that immediately follows the write.

Reply:
The intention of the `xip_address` is to allow bootloaders to find out:
1. What the start address is they need to use,
2. Find out if a copy is being done to the "final" destination, so the bootloader knows it needs to decrypt,

The use that is being described above is NOT the intented use of the `xip_address`.

Reply:
> The intention of the `xip_address` is to allow bootloaders to find out:
> 
>     1. What the start address is they need to use,

I guess ok, though I think that most bootloaders require actual software to have a header that states where the real code begins, and do not check that by device. So no real technical args against, so I guess ok then, though explanation does not convince me.

> 
>     2. Find out if a copy is being done to the "final" destination, so the bootloader knows it needs to decrypt,
> 

That is very vague use case, wouldn't it be better to just mark the storage as "secure" or "non-secure" or under read-out protection? 

> 
> The use that is being described above is NOT the intented use of the `xip_address`.

The use could  land on future enhancements list, because this is improves  access times, though has problems to resolve (I have been evaluating this and have already met some of them).

Reply:
> The use could land on future enhancements list, because this is improves access times, though has problems to resolve (I have been evaluating this and have already met some of them).

The existence of a `xip_address` does not mean that a underlying device is in `xip-mode`. If this would be on the future enhancement list differentiating between `xip-mode` and `non xip-mode` should be introduced. This could be a property bit.